File: engine/python/fife/extensions/pychan/widgets/resizablewindow.py
# -*- coding: utf-8 -*-

# ####################################################################
#  Copyright (C) 2005-2013 by the FIFE team
#  http://www.fifengine.net
#  This file is part of FIFE.
#
#  FIFE is free software; you can redistribute it and/or
#  modify it under the terms of the GNU Lesser General Public
#  License as published by the Free Software Foundation; either
#  version 2.1 of the License, or (at your option) any later version.
#
#  This library is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
#  Lesser General Public License for more details.
#
#  You should have received a copy of the GNU Lesser General Public
#  License along with this library; if not, write to the
#  Free Software Foundation, Inc.,
#  51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
# ####################################################################

from fife import fife
from fife import fifechan

from fife.extensions.pychan.attrs import BoolAttr, MixedListAttr
from common import get_manager

from containers import Window

class ResizableWindow(Window):
	"""
	This is a window class that can be resizable.

	New Attributes
	==============

	  - resizable: Allows resizing with the mouse. If disabled all the side parameters are ignored.
	  - top_resizable: If true, the window can be resized at the top side.
	  - right_resizable: If true, the window can be resized at the right side.
	  - bottom_resizable: If true, the window can be resized at the bottom side.
	  - left_resizable: If true, the window can be resized at the left side.
	  - shove: That allows pull and push in case the min/max size is reached.
	  - cursors: List that contains the cursor definitions. Left, Right, Top, Bottom, LeftTop, RightTop, LeftBottom, RightBottom
	"""

	DEFAULT_RESIZABLE = True
	DEFAULT_TOP_RESIZABLE = True
	DEFAULT_RIGHT_RESIZABLE = True
	DEFAULT_BOTTOM_RESIZABLE = True
	DEFAULT_LEFT_RESIZABLE = True
	DEFAULT_SHOVE = False
	
	
	ATTRIBUTES = Window.ATTRIBUTES + [ BoolAttr('resizable'),
									   BoolAttr('top_resizable'),
									   BoolAttr('right_resizable'),
									   BoolAttr('bottom_resizable'),
									   BoolAttr('left_resizable'),
									   BoolAttr('shove'),
									   MixedListAttr('cursors')
									 ]

	DIRECTION_LIST = [ fifechan.ResizableWindow.CURSOR_DIRECTION_L,
					   fifechan.ResizableWindow.CURSOR_DIRECTION_R,
					   fifechan.ResizableWindow.CURSOR_DIRECTION_T,
					   fifechan.ResizableWindow.CURSOR_DIRECTION_B,
					   fifechan.ResizableWindow.CURSOR_DIRECTION_LT,
					   fifechan.ResizableWindow.CURSOR_DIRECTION_RT,
					   fifechan.ResizableWindow.CURSOR_DIRECTION_LB,
					   fifechan.ResizableWindow.CURSOR_DIRECTION_RB
					 ]

	def __init__(self, 
				 parent = None,
				 name = None,
				 size = None,
				 min_size = None,
				 max_size = None,
				 fixed_size = None,
				 margins = None,
				 padding = None,
				 helptext = None,
				 position = None,
				 style = None,
				 hexpand = None,
				 vexpand = None,
				 font = None,
				 base_color = None,
				 background_color = None,
				 foreground_color = None,
				 selection_color = None,
				 border_color = None,
				 outline_color = None,
				 border_size = None,
				 outline_size = None,
				 position_technique = None,
				 is_focusable = None,
				 comment = None,
				 background_image = None,
				 opaque = None,
				 layout = None,
				 spacing = None,
				 uniform_size = None,
				 _real_widget = None,
				 title = None,
				 titlebar_height = None,
				 movable = None,
				 resizable = None,
				 top_resizable = None,
				 right_resizable = None,
				 bottom_resizable = None,
				 left_resizable = None,
				 shove = None,
				 cursors = None):
		
		if _real_widget is None: _real_widget = fifechan.ResizableWindow()
			
		super(ResizableWindow,self).__init__(parent=parent,
											 name=name,
											 size=size,
											 min_size=min_size,
											 max_size=max_size,
											 fixed_size=fixed_size,
											 margins=margins,
											 padding=padding,
											 helptext=helptext,
											 position=position,
											 style=style,
											 hexpand=hexpand,
											 vexpand=vexpand,
											 font=font,
											 base_color=base_color,
											 background_color=background_color,
											 foreground_color=foreground_color,
											 selection_color=selection_color,
											 border_color=border_color,
											 outline_color=outline_color,
											 border_size=border_size,
											 outline_size=outline_size,
											 position_technique=position_technique,
											 is_focusable=is_focusable,
											 comment=comment,
											 background_image=background_image,
											 opaque=opaque,
											 layout=layout,
											 spacing=spacing,
											 uniform_size=uniform_size,
											 _real_widget=_real_widget,
											 title=title,
											 titlebar_height=titlebar_height,
											 movable=movable)

		if resizable is not None: self.resizable = resizable
		else: self.resizable = self.DEFAULT_RESIZABLE

		if top_resizable is not None: self.top_resizable = top_resizable
		else: self.top_resizable = self.DEFAULT_TOP_RESIZABLE

		if right_resizable is not None: self.right_resizable = right_resizable
		else: self.right_resizable = self.DEFAULT_RIGHT_RESIZABLE

		if bottom_resizable is not None: self.bottom_resizable = bottom_resizable
		else: self.bottom_resizable = self.DEFAULT_BOTTOM_RESIZABLE

		if left_resizable is not None: self.left_resizable = left_resizable
		else: self.left_resizable = self.DEFAULT_LEFT_RESIZABLE

		if shove is not None: self.shove = shove
		else: self.shove = self.DEFAULT_SHOVE

		if cursors is not None: self.cursors = cursors

	def clone(self, prefix):
		windowClone = ResizableWindow(None, 
					self._createNameWithPrefix(prefix),
					self.size,
					self.min_size, 
					self.max_size,
					self.fixed_size,
					self.margins,
					self.padding,
					self.helptext, 
					self.position, 
					self.style, 
					self.hexpand,
					self.vexpand,
					self.font,
					self.base_color,
					self.background_color,
					self.foreground_color,
					self.selection_color,
					self.border_color,
					self.outline_color,
					self.border_size,
					self.outline_size,
					self.position_technique,
					self.is_focusable,
					self.comment,
					self.background_image,
					self.opaque,
					self.layout,
					self.spacing,
					self.uniform_size,
					None,
					self.title,
					self.titlebar_height,
					self.movable,
					self.resizable,
					self.top_resizable,
					self.right_resizable,
					self.bottom_resizable,
					self.left_resizable,
					self.shove,
					self.cursors)
		
		windowClone.addChildren(self._cloneChildren(prefix))		     
		return windowClone

	def _setResizable(self, resizable): self.real_widget.setResizable(resizable)
	def _getResizable(self): return self.real_widget.getResizable()
	resizable = property(_getResizable, _setResizable)

	def _setTopResizable(self, resizable): self.real_widget.setTopResizable(resizable)
	def _getTopResizable(self): return self.real_widget.getTopResizable()
	top_resizable = property(_getTopResizable, _setTopResizable)

	def _setRightResizable(self, resizable): self.real_widget.setRightResizable(resizable)
	def _getRightResizable(self): return self.real_widget.getRightResizable()
	right_resizable = property(_getRightResizable, _setRightResizable)

	def _setBottomResizable(self, resizable): self.real_widget.setBottomResizable(resizable)
	def _getBottomResizable(self): return self.real_widget.getBottomResizable()
	bottom_resizable = property(_getBottomResizable, _setBottomResizable)

	def _setLeftResizable(self, resizable): self.real_widget.setLeftResizable(resizable)
	def _getLeftResizable(self): return self.real_widget.getLeftResizable()
	left_resizable = property(_getLeftResizable, _setLeftResizable)

	def _setShove(self, shove): self.real_widget.setShove(shove)
	def _getShove(self): return self.real_widget.getShove()
	shove = property(_getShove, _setShove)

	def _setCursor(self, direction, cursor):
		self.real_widget.set(direction, cursor)
	def _getCursor(self, direction):
		type = self.real_widget.getType(direction)
		if type is fife.CURSOR_NATIVE:
			return self.real_widget.getId(direction)
		elif type is fife.CURSOR_IMAGE:
			return self.real_widget.getImage(direction)
		elif type is fife.CURSOR_Animation:
			return self.real_widget.getAnimation(direction)
	cursor = property(_getCursor, _setCursor)

	def _setCursors(self, cursors):
		if cursors is not None:
			i = 0
			for c in cursors:
				if isinstance(c, str):
					# we need a fife image
					c = get_manager().loadImage(c, gui=False)
				self.real_widget.set(self.DIRECTION_LIST[i], c)
				i += 1
	def _getCursors(self):
		cursors = []
		for d in self.DIRECTION_LIST:
			cursors.append(self.cursor(d))
		return cursors
	cursors = property(_getCursors, _setCursors)

File: engine/core/gui/fifechan/widgets/resizablewindow.cpp
/***************************************************************************
 *   Copyright (C) 2005-2013 by the FIFE team                              *
 *   http://www.fifengine.net                                              *
 *   This file is part of FIFE::                                            *
 *                                                                         *
 *   FIFE is free software; you can redistribute it and/or                 *
 *   modify it under the terms of the GNU Lesser General Public            *
 *   License as published by the Free Software Foundation; either          *
 *   version 2.1 of the License, or (at your option) any later version.    *
 *                                                                         *
 *   This library is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU     *
 *   Lesser General Public License for more details.                       *
 *                                                                         *
 *   You should have received a copy of the GNU Lesser General Public      *
 *   License along with this library; if not, write to the                 *
 *   Free Software Foundation, Inc.,                                       *
 *   51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA          *
 ***************************************************************************/

// Standard C++ library includes
#include <cassert>

// 3rd party library includes

// FIFE includes
// These includes are split up in two parts, separated by one empty line
// First block: files included from the FIFE root src directory
// Second block: files included from the same folder
#include "gui/fifechan/base/gui_font.h"
#include "gui/fifechan/fifechanmanager.h"
#include "util/base/exception.h"

#include "resizablewindow.h"

namespace fcn {
	ResizableWindow::ResizableWindow():
		Window(),
		m_cursor(FIFE::FifechanManager::instance()->getCursor()),
		m_borderDistance(5),
		m_resizable(true),
		m_resizableTop(true),
		m_resizableRight(true),
		m_resizableBottom(true),
		m_resizableLeft(true),
		m_resizing(false),
		m_resizeTop(false),
		m_resizeRight(false),
		m_resizeBottom(false),
		m_resizeLeft(false),
		m_shove(false) {

		addKeyListener(this);
		addFocusListener(this);
		initCursors();
	}

	ResizableWindow::ResizableWindow(bool resizable):
		Window(),
		m_cursor(FIFE::FifechanManager::instance()->getCursor()),
		m_borderDistance(5),
		m_resizable(resizable),
		m_resizableTop(true),
		m_resizableRight(true),
		m_resizableBottom(true),
		m_resizableLeft(true),
		m_resizing(false),
		m_resizeTop(false),
		m_resizeRight(false),
		m_resizeBottom(false),
		m_resizeLeft(false),
		m_shove(false) {

		addKeyListener(this);
		addFocusListener(this);
		initCursors();
	}

	ResizableWindow::ResizableWindow(const std::string& caption, bool resizable):
		Window(),
		m_cursor(FIFE::FifechanManager::instance()->getCursor()),
		m_borderDistance(5),
		m_resizable(resizable),
		m_resizableTop(true),
		m_resizableRight(true),
		m_resizableBottom(true),
		m_resizableLeft(true),
		m_resizing(false),
		m_resizeTop(false),
		m_resizeRight(false),
		m_resizeBottom(false),
		m_resizeLeft(false),
		m_shove(false) {

		setCaption(caption);
		addKeyListener(this);
		addFocusListener(this);
		initCursors();
	}

	ResizableWindow::~ResizableWindow() {
	}
	
	void ResizableWindow::initCursors() {
		// defines default cursors
		// don't change the sequence!
		CursorState state;
		state.cursorId = FIFE::NC_ARROW;
		state.cursorType = FIFE::CURSOR_NONE;
		m_saved = state;
		state.cursorType = FIFE::CURSOR_NATIVE;
		// left
		state.cursorId = FIFE::NC_RESIZEWE;
		m_cursors.push_back(state);
		// right
		state.cursorId = FIFE::NC_RESIZEWE;
		m_cursors.push_back(state);
		// top
		state.cursorId = FIFE::NC_RESIZENS;
		m_cursors.push_back(state);
		// left and top
		state.cursorId = FIFE::NC_RESIZENWSE;
		m_cursors.push_back(state);
		// right and top
		state.cursorId = FIFE::NC_RESIZENESW;
		m_cursors.push_back(state);
		// bottom
		state.cursorId = FIFE::NC_RESIZENS;
		m_cursors.push_back(state);
		// left and bottom
		state.cursorId = FIFE::NC_RESIZENESW;
		m_cursors.push_back(state);
		// right and bottom
		state.cursorId = FIFE::NC_RESIZENWSE;
		m_cursors.push_back(state);
	}

	void ResizableWindow::restoreCursor() {
		FIFE::MouseCursorType type = m_saved.cursorType;
		switch (type) {
			case FIFE::CURSOR_NATIVE:
				m_cursor->set(m_saved.cursorId);
				break;
			case FIFE::CURSOR_IMAGE:
				m_cursor->set(m_saved.cursorImage);
				break;
			case FIFE::CURSOR_ANIMATION:
				m_cursor->set(m_saved.cursorAnimation);
				break;
			default:
				return;
		}
	}

	void ResizableWindow::saveCursor() {
		FIFE::MouseCursorType type = m_cursor->getType();
		switch (type) {
			case FIFE::CURSOR_NATIVE:
				m_saved.cursorType = type;
				m_saved.cursorId = m_cursor->getId();
				m_saved.cursorImage.reset();
				m_saved.cursorAnimation.reset();
				break;
			case FIFE::CURSOR_IMAGE:
				m_saved.cursorType = type;
				m_saved.cursorId = FIFE::NC_ARROW;;
				m_saved.cursorImage = m_cursor->getImage();
				m_saved.cursorAnimation.reset();
				break;
			case FIFE::CURSOR_ANIMATION:
				m_saved.cursorType = type;
				m_saved.cursorId = FIFE::NC_ARROW;;
				m_saved.cursorImage.reset();
				m_saved.cursorAnimation = m_cursor->getAnimation();
				break;
			default:
				m_saved.cursorType = FIFE::CURSOR_NONE;
		}
	}

	void ResizableWindow::setResizableBorderDistance(int32_t border) {
		m_borderDistance = border;
	}

	int32_t ResizableWindow::getResizableBorderDistance() const {
		return m_borderDistance;
	}

	void ResizableWindow::setResizable(bool resizable) {
		m_resizable = resizable;
	}

	bool ResizableWindow::isResizable() const {
		return m_resizable;
	}

	void ResizableWindow::setTopResizable(bool resizable) {
		m_resizableTop = resizable;
	}

	bool ResizableWindow::isTopResizable() const {
		return m_resizableTop;
	}

	void ResizableWindow::setRightResizable(bool resizable) {
		m_resizableRight = resizable;
	}

	bool ResizableWindow::isRightResizable() const {
		return m_resizableRight;
	}

	void ResizableWindow::setBottomResizable(bool resizable) {
		m_resizableBottom = resizable;
	}

	bool ResizableWindow::isBottomResizable() const {
		return m_resizableBottom;
	}

	void ResizableWindow::setLeftResizable(bool resizable) {
		m_resizableLeft = resizable;
	}

	bool ResizableWindow::isLeftResizable() const {
		return m_resizableLeft;
	}

	void ResizableWindow::setShove(bool shove) {
		m_shove = shove;
	}

	bool ResizableWindow::getShove() const {
		return m_shove;
	}

	void ResizableWindow::set(CursorDirections direction, uint32_t cursor_id) {
		CursorState& state = m_cursors[direction];
		state.cursorType = FIFE::CURSOR_NATIVE;
		state.cursorId = cursor_id;
		state.cursorImage.reset();
		state.cursorAnimation.reset();
	}

	void ResizableWindow::set(CursorDirections direction, FIFE::ImagePtr image) {
		CursorState& state = m_cursors[direction];
		state.cursorType = FIFE::CURSOR_IMAGE;
		state.cursorId = FIFE::NC_ARROW;
		state.cursorImage = image;
		state.cursorAnimation.reset();
	}

	void ResizableWindow::set(CursorDirections direction, FIFE::AnimationPtr anim) {
		CursorState& state = m_cursors[direction];
		state.cursorType = FIFE::CURSOR_ANIMATION;
		state.cursorId = FIFE::NC_ARROW;
		state.cursorImage.reset();
		state.cursorAnimation = anim;
	}

	FIFE::MouseCursorType ResizableWindow::getType(CursorDirections direction) const {
		return m_cursors[direction].cursorType;
	}
	
	uint32_t ResizableWindow::getId(CursorDirections direction) const {
		return m_cursors[direction].cursorId;
	}

	FIFE::ImagePtr ResizableWindow::getImage(CursorDirections direction) {
		return m_cursors[direction].cursorImage;
	}

	FIFE::AnimationPtr ResizableWindow::getAnimation(CursorDirections direction) {
		return m_cursors[direction].cursorAnimation;
	}

	void ResizableWindow::resizeToContent(bool recursiv) {
		// we keep the size in case we want it resizable
		if (m_resizable) {
			int w = getWidth();
			int h = getHeight();
			Window::resizeToContent(recursiv);
			setSize(w, h);
		} else {
			Window::resizeToContent(recursiv);
		}
	}

	void ResizableWindow::mouseEntered(MouseEvent& mouseEvent) {
		if (m_resizable && !m_resizing) {
			saveCursor();
		}
		Window::mouseEntered(mouseEvent);
	}

	void ResizableWindow::mouseExited(MouseEvent& mouseEvent) {
		if (m_resizable && !m_resizing && m_saved.cursorType != FIFE::CURSOR_NONE) {
			restoreCursor();
		} else if (m_resizable && m_resizing) {
			mouseEvent.consume();
		}
		Window::mouseExited(mouseEvent);
	}

	void ResizableWindow::mousePressed(MouseEvent& mouseEvent) {
		if (m_resizable && mouseEvent.getButton() == MouseEvent::Left) {
			m_resizeLeft = m_resizableLeft && mouseEvent.getX() < m_borderDistance;
			m_resizeRight = m_resizableRight && mouseEvent.getX() > getWidth() - m_borderDistance;
			m_resizeTop = m_resizableTop && mouseEvent.getY() < m_borderDistance;
			m_resizeBottom = m_resizableBottom && mouseEvent.getY() > getHeight() - m_borderDistance;
			if (m_resizeTop || m_resizeBottom || m_resizeLeft || m_resizeRight) {
				m_resizing = true;
				mouseEvent.consume();
			}
		}
		Window::mousePressed(mouseEvent);
	}

	void ResizableWindow::mouseReleased(MouseEvent& mouseEvent) {
		if (m_resizing) {
			adaptLayout();
			m_resizing = false;
			if (mouseEvent.getX() <= 0 || mouseEvent.getX() >= getWidth() ||
				mouseEvent.getY() <= 0 || mouseEvent.getY() >= getHeight()) {
					mouseExited(mouseEvent);
			}
			mouseEvent.consume();
		} else {
			Window::mouseReleased(mouseEvent);
		}
	}

	void ResizableWindow::mouseMoved(MouseEvent& mouseEvent) {
		if (m_resizable && !m_resizing && m_saved.cursorType != FIFE::CURSOR_NONE) {
			// use the CursorDirections directly
			uint32_t index = 0;
			index += (m_resizableLeft && mouseEvent.getX() < m_borderDistance) ? 1 : 0;
			index += (m_resizableRight && mouseEvent.getX() > getWidth() - m_borderDistance) ? 2 : 0;
			index += (m_resizableTop && mouseEvent.getY() < m_borderDistance) ? 3 : 0;
			index += (m_resizableBottom && mouseEvent.getY() > getHeight() - m_borderDistance) ? 6 : 0;

			if (index > 0) {
				// to get the real index value
				index -= 1;
				CursorState state = m_cursors[index];
				FIFE::MouseCursorType type = state.cursorType;
				switch (type) {
					case FIFE::CURSOR_NATIVE:
						m_cursor->set(state.cursorId);
						break;
					case FIFE::CURSOR_IMAGE:
						m_cursor->set(state.cursorImage);
						break;
					case FIFE::CURSOR_ANIMATION:
						m_cursor->set(state.cursorAnimation);
						break;
					default:
						break;
				}
				mouseEvent.consume();
			} else {
				restoreCursor();
			}
		}
		Window::mouseMoved(mouseEvent);
	}
	
	void ResizableWindow::mouseDragged(MouseEvent& mouseEvent) {
		if (m_resizable && m_resizing) {
			int x = mouseEvent.getX();
			int y = mouseEvent.getY();
			if (m_resizeLeft) {
				int oldW = getWidth();
				int newW = oldW - x;
				if (newW < 0) {
					newW = 0;
				}
				setWidth(newW);
				if (m_shove) {
					setX(getX() + x);
				} else {
					newW = getWidth();
					int newX = oldW - newW;
					setX(getX() + newX);
				}
			} else if (m_resizeRight) {
				setWidth(x);
				if (x != getWidth() && m_shove) {
					setX(getX() + (x-getWidth()));
				}
			}

			if (m_resizeTop) {
				int oldH = getHeight();
				int newH = oldH - y;
				if (newH < 0) {
					newH = 0;
				}
				setHeight(newH);
				if (m_shove) {
					setY(getY() + y);
				} else {
					newH = getHeight();
					int newY = oldH - newH;
					setY(getY() + newY);
				}
			} else if (m_resizeBottom) {
				setHeight(y);
				if (y != getHeight() && m_shove) {
					setY(getY() + (y-getHeight()));
				}
			}
			mouseEvent.consume();
		} else {
			Window::mouseDragged(mouseEvent);
		}
	}


	void ResizableWindow::focusLost(const Event& event) {
		m_resizing = false;
		restoreCursor();
	}
}
